tapsend/proof.go

```diff
@@ -326,9 +339,86 @@ func proofParams(finalTx *wire.MsgTx, vPkt *tappsbt.VPacket,
 	return splitParams, nil
 }
 
+// addSTXOExclusionProofs adds exclusion proofs for all the STXOs of the asset,
+// for all the outputs that are asset outputs but haven't been processed yet,
+// otherwise they'll be skipped. This should only be called after
+// `addOtherOutputExclusionProofs` because it depends on
+// `params.ExclusionProofs` already being set.
+func addSTXOExclusionProofs(outputs []*tappsbt.VOutput,
+	newAsset *asset.Asset, params *proof.TransitionParams,
+	outputCommitments map[uint32]*commitment.TapCommitment) error {
+
+	stxoAssets, err := asset.CollectSTXO(newAsset)
+	if err != nil {
+		return fmt.Errorf("error collecting STXO assets: %w", err)
+	}
+
+	for idx := range outputs {
+		vOut := outputs[idx]
+
+		outIndex := vOut.AnchorOutputIndex
+		// We can use `HaveInclusionProof` here because it is just a
+		// check on whether we are processing our own anchor output.
+		haveIProof := params.HaveInclusionProof(outIndex)
+		haveEProof := params.HaveSTXOExclusionProof(outIndex)
+		if haveIProof || haveEProof {
+			continue
+		}
+
+		tapTree := outputCommitments[outIndex]
+
+		for idx := range stxoAssets {
+			stxoAsset := stxoAssets[idx].(*asset.Asset)
+			pubKey := stxoAsset.ScriptKey.PubKey
+			identifier := asset.ToSerialized(pubKey)
+
+			_, exclusionProof, err := tapTree.Proof(
+				stxoAsset.TapCommitmentKey(),
+				stxoAsset.AssetCommitmentKey(),
+			)
+			if err != nil {
+				return err
+			}
+
```

Comment:
Perhaps we should have a sanity check that this is actually an exclusion proof as a sanity check? I think we can rely on the empty hash abstraction here.

Reply:
I think I can do one better.

I've added code some 20 lines below that confirms that we are creating the stxo proofs for the asset for which there already is an (old style) exclusion proof. So this confirms that this is actually an exclusion proof, but also confirms that it's a valid exclusion proof for this exact asset.